system/lib/js_math.cpp

```diff
@@ -0,0 +1,40 @@
+#include <emscripten.h>
+#include <math.h>
+#include <stdlib.h>
+
+#define CALL_JS_1(cname, jsname, type, casttype) \
+  EM_JS(type, JS_##cname, (type x), { return jsname(x) }); \
+  type cname(type x) { return JS_##cname((casttype)x); }
+
+#define CALL_JS_1_TRIPLE(cname, jsname) \
+  CALL_JS_1(cname, jsname, double, double) \
+  CALL_JS_1(cname##f, jsname, float, float) \
+  CALL_JS_1(cname##l, jsname, long double, double)
+
+#define CALL_JS_2(cname, jsname, type, casttype) \
+  EM_JS(type, JS_##cname, (type x, type y), { return jsname(x, y) }); \
+  type cname(type x, type y) { return JS_##cname((casttype)x, (casttype)y); }
+
+#define CALL_JS_2_TRIPLE(cname, jsname) \
+  CALL_JS_2(cname, jsname, double, double) \
+  CALL_JS_2(cname##f, jsname, float, float) \
+  CALL_JS_2(cname##l, jsname, long double, double)
+
+CALL_JS_1_TRIPLE(cos, Math.cos)
+CALL_JS_1_TRIPLE(sin, Math.sin)
+CALL_JS_1_TRIPLE(tan, Math.tan)
+CALL_JS_1_TRIPLE(acos, Math.acos)
+CALL_JS_1_TRIPLE(asin, Math.asin)
+CALL_JS_1_TRIPLE(atan, Math.atan)
+CALL_JS_2_TRIPLE(atan2, Math.atan2)
+CALL_JS_1_TRIPLE(exp, Math.exp)
+CALL_JS_1_TRIPLE(log, Math.log)
+CALL_JS_1_TRIPLE(sqrt, Math.sqrt)
+CALL_JS_1_TRIPLE(fabs, Math.abs)
+CALL_JS_1_TRIPLE(ceil, Math.ceil)
+CALL_JS_1_TRIPLE(floor, Math.floor)
+CALL_JS_2_TRIPLE(pow, Math.pow)
+CALL_JS_1_TRIPLE(round, Math.round)
+CALL_JS_1_TRIPLE(rint, Math.round)
+CALL_JS_1_TRIPLE(nearbyint, Math.round)
```

Comment:
Not sure if relevant but iirc JS has strange semantics here, with `Math.round` unconditionally rounding towards +Infinity, while C rounds towards zero.

Reply:
pow will probably need some special casing too: #4980

Reply:
Yeah. pow in C++ and JS [slightly different for Infinity and NaN](https://github.com/AssemblyScript/assemblyscript/blob/master/tests/compiler/std/math.ts#L2635)

Reply:
Yeah, see the documentation in `settings.js`, these differences are documented, and also that results can vary between browsers.

I'm not sure it makes sense to have a more accurate version. The more accurate one would be larger, and it would still not be identical to libc (still vary between browsers etc.).

Reply:
Hmm, which browser not follow this? Because it relate to test262 tests
https://github.com/tc39/test262/blob/ee3715ee56744ccc8aeb22a921f442e98090b3c1/test/built-ins/Math/pow/applying-the-exp-operator_A4.js

https://github.com/tc39/test262/blob/ee3715ee56744ccc8aeb22a921f442e98090b3c1/test/built-ins/Math/pow/applying-the-exp-operator_A1.js

Reply:
These differences are part of the ES standard, we can rely on them being consistent across browsers.

Maybe if they're not handled, could the settings.js comment describe them in more detail? At least for the differences that are known.

Reply:
Sorry if I wasn't clear. Yes, there are two sets of differences. First, the spec of libc and JS are different. Browsers do follow the JS spec. We can try to make these functions behave more like the libc one, if we think it's worth it.

The other set of differences are that the JS spec doesn't specify every bit in the output. There have historically been many differences between them, including ones that browsers made and had to revert despite them being within what the spec allows. My info may be out of date here, though, I haven't followed this recently.

Reply:
I fixed up rint and nearbyint, which should handle the rounding. Seems easy enough.

Corner cases in infinity etc. seem excessive to me personally...